overlord/hookstate/ctlcmd/fde_setup.go

```diff
@@ -0,0 +1,143 @@
+$ snapctl fde-setup-request
+{"op":"features"}
+$ echo '[]' | snapctl fde-setup-result
+
+Alternatively the hook could reply with:
+$ echo '{"error":"hardware-unsupported"}' | snapctl fde-setup-result
+
+And then it is called again with a request to do the initial key setup:
+$ snapctl fde-setup-request
+{"op":"initial-setup", "key": "key-to-seal", "key-name":"key-for-ubuntu-data"}
+$ echo "$sealed_key" | snapctl fde-setup-result
+`)
+
+func init() {
+	addCommand("fde-setup-request", shortFdeSetupRequestHelp, longFdeSetupRequestHelp, func() command { return &fdeSetupRequestCommand{} })
+}
+
+func (c *fdeSetupRequestCommand) Execute(args []string) error {
+	context := c.context()
+	if context == nil {
+		return fmt.Errorf("cannot run fde-setup-request without a context")
+	}
+	context.Lock()
+	defer context.Unlock()
+
+	if context.HookName() != "fde-setup" {
+		return fmt.Errorf("cannot use fde-setup-request outside of the fde-setup hook")
+	}
+
+	var fdeSetup hookstate.FDESetupRequest
+	if err := context.Get("fde-setup-request", &fdeSetup); err != nil {
+		return fmt.Errorf("cannot get fde-setup-op from context: %v", err)
+	}
+	// Op is either "initial-setup" or "features"
+	switch fdeSetup.Op {
+	case "features", "initial-setup":
+		// fine
+	default:
+		return fmt.Errorf("unknown fde-setup-request op %q", fdeSetup.Op)
+
+	}
+
+	bytes, err := json.Marshal(fdeSetup)
+	if err != nil {
+		return fmt.Errorf("cannot json print fde key: %v", err)
+	}
+	c.printf("%s\n", string(bytes))
+
+	return nil
+}
+
+type fdeSetupResultCommand struct {
+	baseCommand
+}
+
+var shortFdeSetupResultHelp = i18n.G("Set result for full disk encryption")
+var longFdeSetupResultHelp = i18n.G(`
+The fde-setup-result command sets the result data for a fde-setup hook
+reading it from stdin.
+
+E.g.
+When the fde-setup hook is called with "op":"features:
+$ echo "[]" | snapctl fde-setup-result
+
+Or when the fde-setup hook is called with "op":"initial-setup":
+$ echo "sealed-key" | snapctl fde-setup-result
+`)
+
+func init() {
+	addCommand("fde-setup-result", shortFdeSetupResultHelp, longFdeSetupResultHelp, func() command { return &fdeSetupResultCommand{} })
+}
+
+func (c *fdeSetupResultCommand) Execute(args []string) error {
+	context := c.context()
+	if context == nil {
+		return fmt.Errorf("cannot run fde-setup-result without a context")
+	}
+	context.Lock()
+	defer context.Unlock()
+
+	if context.HookName() != "fde-setup" {
+		return fmt.Errorf("cannot use fde-setup-result outside of the fde-setup hook")
+	}
+
+	var fdeSetupResult []byte
+	if err := context.Get("stdin", &fdeSetupResult); err != nil {
+		return fmt.Errorf("internal error: cannot get result from stdin: %v", err)
+	}
+	if fdeSetupResult == nil {
+		return fmt.Errorf("no result data found from stdin")
+	}
+	task, ok := context.Task()
+	if !ok {
+		return fmt.Errorf("internal error: fdeSetupResultCommand called without task")
+	}
+	task.Set("fde-setup-result", fdeSetupResult)
```

Comment:
can you explain why we are setting it on the task? we could set it on the context and retrieve it as needed using a OnDone callback on the context maybe? I suppose it's clearer from the runner but the runner will come later

Reply:
Aha, I did not think about using OnDone() for this, I changed this code now to set it on the context and will change the runner too.